File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */
    lowestCommonAncestor(
        root: TreeNode | null,
        p: TreeNode | null,
        q: TreeNode | null,
    ) {

        while (root) {
            if (root.val == p.val || root.val == q.val) {
                return root
            }

            else if ((p.val < root.val && q.val > root.val) || (p.val > root.val && q.val < root.val)) {
                return root
            }

            else if (p.val > root.val && q.val > root.val) {
                root = root.right
            }

            else {
                root = root.left
            }
        }

    }
}

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:

        res = []

        if not root:
            return res

        queue = collections.deque()
        queue.append((root,0))

        level = 0

        while queue:

            while len(queue) > 0 and queue[0][1] == level:
                node, i = queue.popleft()

                if node.left:
                    queue.append((node.left, level+1))

                if node.right:
                    queue.append((node.right, level+1))

            res.append(node.val)
            level += 1

        return res

            


        